# iUma22: Genome scale metabolic model for Ustilago maydis

*Ustilago maydis* is an important plant pathogen causing corn-smut disease and an effective biotechnological production host. The lack of a comprehensive metabolic overview hinders a full understanding of environmental adaptation and a full use of the organismâ€™s metabolic potential. Here, we report the first genome scale metabolic model (GSMM) of *Ustilago maydis* (iUma22) for the simulation of metabolic activities. iUma22 was reconstructed from sequencing and annotation using PathwayTools, the biomass equation was derived from literature values and from the codon composition. The final model contains over 25% of annotated genes in the sequenced genome. Substrate utilization was corrected by Biolog-Phenotype arrays and exponential batch cultivations were used to test growth predictions. A pan-genome of four different *U. maydis* strains revealed missing metabolic pathways in iUma22. The majority of metabolic differences between iUma22 and the pan-genome occurs in the inositol, purine and starch metabolic pathways. The new model allows studies of metabolic adaptations to different environmental niches as well as for biotechnological applications.

## **Citation**

Liebal et al. (2022), A Genome-Scale Metabolic Model for the Smut-Fungus *Ustilago maydis*. bioRxiv, doi:



## Keywords

**Utilisation:** experimental data reconstruction

**Field:** metabolic-network reconstruction

**Type of model:** curated

**Model source:** PathwayTools

**Taxonomic name:** Ustilago maydis

**Strain:** 521

**Metabolic system: **general metabolism

**Condition: **aerobic; glucose limited; glycerol limited; defined media

## Usage

Online simulation via Binder:
[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/iAMB-RWTH-Aachen/Ustilago_maydis-GEM/master)

## Contact

[Ulf Liebal](ulf.liebal@iamb.rwth-aachen.de)
