{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genome scale model reconstruction of the smut fungus *Ustilago maydis*: Reproduction of simulations with CobraPy Notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction\n",
    "\n",
    "This notebook is an extension to the article by Liebal et al., (2022) and reproduces the comparison of experimental growth rates with predictions by the genome scale metabolic model (GSMM) of *Ustilago maydis*. Finaly, the Memote quality check is performed. \n",
    "\n",
    "## Requirements\n",
    "Make sure that the required files are present in the notebook directory. \n",
    "\n",
    "Files:\n",
    " * ../model/iCL1079.xml: SBML file, Genome scale metabolic model of *U. maydis*\n",
    " \n",
    "Dependencies:\n",
    " * CPython 3.7.6\n",
    " * IPython 7.12.0\n",
    " * ipywidgets 7.5.1\n",
    " * matplotlib 3.1.3\n",
    " * numpy 1.18.1\n",
    " * pandas 1.0.1\n",
    " * cobra 0.17.1\n",
    " * memote 0.10.2\n",
    "\n",
    "## Set-up compute environment \n",
    "\n",
    "In the following cell the python environment is set up. If CobraPy and Memote are not already downloaded, use the following command to download the packages them from the cell:\n",
    "!{sys.executable} -m pip install cobra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "System initiated.\n"
     ]
    }
   ],
   "source": [
    "import sys # loading commands to control/navigate within the system architecture\n",
    "# Loading pandas, a library for data manipulation\n",
    "import os\n",
    "# import xlrd\n",
    "import pandas as pd\n",
    "# import lxml\n",
    "# for export of biolog substrate test dictionary\n",
    "import csv\n",
    "import itertools\n",
    "import operator\n",
    "\n",
    "\n",
    "# Loading numpy, a library fo manipulation of numbers\n",
    "import numpy as np\n",
    "\n",
    "# loading matplotlib, a library for visualization\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "# loading cobrapy, a library dedicated to the analysis of genome scale metabolic models\n",
    "from cobra.io import read_sbml_model, write_sbml_model\n",
    "from cobra import Reaction, Metabolite\n",
    "\n",
    "# from escher import Builder\n",
    "\n",
    "# loading Memote, quality assessment of GSMM\n",
    "# from memote import test_model, snapshot_report\n",
    "# from FastaTools import *\n",
    "from iambcodes.cobra import *\n",
    "from iambcodes.fasta import *\n",
    "\n",
    "print('System initiated.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Model loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>UMAY</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07f32e833d430</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>1230</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>1859</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*BIOMASS_REACTION - 1.0*BIOMASS_REACTION_reverse_42726</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>Cytoplasm, Extracellular space, Mitochondrial Lumen, Mitochondrial Intermembrane Space</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model UMAY at 0x7f32e833d430>"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ModelFile = os.path.join('..','model', 'iUma22_Curated.xml') # iCL1079.xml 'iUma22_Curated.xml'\n",
    "model=read_sbml_model(ModelFile)\n",
    "model\n",
    "# model.metabolites"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Support functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def TestSubstrate(model, EX_Sub, Rate=10):\n",
    "    '''\n",
    "    Function for testing substrate consumption with exchange reaction.\n",
    "    '''\n",
    "    with model:\n",
    "        medium = model.medium\n",
    "        medium[EX_Sub] = Rate\n",
    "        medium['EX_co2_e'] = 0\n",
    "        medium['EX_glc__D_e'] = 0\n",
    "        model.medium = medium\n",
    "        return round(model.slim_optimize(),2)\n",
    "\n",
    "def MetNewFromExist(MetIn):\n",
    "    Metout = Metabolite('{}_e'.format(MetIn.id[:-2]),\n",
    "                       formula = MetIn.formula,\n",
    "                       name = MetIn.name,\n",
    "                       compartment = 'e')\n",
    "    return Metout\n",
    "\n",
    "def CreateTransReact(MetIn):\n",
    "    reaction = Reaction('Trans_{}'.format(MetIn.id),\n",
    "                       name = 'Transport c<->e {}'.format(MetIn.name),\n",
    "                       lower_bound = -1000,\n",
    "                       upper_bound = 1000)\n",
    "    MetOut = MetNewFromExist(MetIn)\n",
    "    reaction.add_metabolites({MetIn:-1.0, MetOut:1.0})\n",
    "    \n",
    "    return reaction\n",
    "\n",
    "def CSVexport(myDict, FName='BiologGrowthTest'):\n",
    "    fields = ['id','name','id_e','id_c','growth','CL_Growth']\n",
    "    with open(FName, 'w') as f:\n",
    "        w = csv.DictWriter(f, fields)\n",
    "        w.writeheader()\n",
    "        for k in myDict:\n",
    "            w.writerow({field: myDict[k].get(field) or k for field in fields})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data loading\n",
    "\n",
    "Now, the external data is integrated. The excel-sheet is parsed and the GSMM file is converted to a model variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ulf/anaconda3/envs/iambjudas/lib/python3.9/site-packages/openpyxl/worksheet/_reader.py:312: UserWarning: Unknown extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Biolog_Name</th>\n",
       "      <th>Andere Namen</th>\n",
       "      <th>BiGG_ID</th>\n",
       "      <th>growth_experiment</th>\n",
       "      <th>In_Silico_Growth_Rate</th>\n",
       "      <th>Growth_Prediction</th>\n",
       "      <th>Boundary_Present</th>\n",
       "      <th>Status</th>\n",
       "      <th>Confusion_Marker</th>\n",
       "      <th>Unnamed: 10</th>\n",
       "      <th>Transporter/Gapfilling/renaming/pathway unclear</th>\n",
       "      <th>comments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>29</td>\n",
       "      <td>D-Ribose</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rib__D</td>\n",
       "      <td>1</td>\n",
       "      <td>0.250015</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>In model</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>76</td>\n",
       "      <td>Pyruvic Acid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pyr</td>\n",
       "      <td>1</td>\n",
       "      <td>0.086929</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>In model</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>32</td>\n",
       "      <td>D-Xylose</td>\n",
       "      <td>NaN</td>\n",
       "      <td>xyl__D</td>\n",
       "      <td>1</td>\n",
       "      <td>0.433506</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>In model</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0   Biolog_Name Andere Namen BiGG_ID  growth_experiment  \\\n",
       "0          29      D-Ribose          NaN  rib__D                  1   \n",
       "1          76  Pyruvic Acid          NaN     pyr                  1   \n",
       "2          32      D-Xylose          NaN  xyl__D                  1   \n",
       "\n",
       "   In_Silico_Growth_Rate  Growth_Prediction  Boundary_Present    Status  \\\n",
       "0               0.250015               True              True  In model   \n",
       "1               0.086929               True              True  In model   \n",
       "2               0.433506               True              True  In model   \n",
       "\n",
       "   Confusion_Marker Unnamed: 10  \\\n",
       "0                 1         NaN   \n",
       "1                 1         NaN   \n",
       "2                 1         NaN   \n",
       "\n",
       "  Transporter/Gapfilling/renaming/pathway unclear comments  \n",
       "0                                             NaN      NaN  \n",
       "1                                             NaN      NaN  \n",
       "2                                             NaN      NaN  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myxls = os.path.join('..','data','MSc-Lieven_Christian','sorted_growth_study.xlsx')\n",
    "x1 = pd.ExcelFile(myxls)\n",
    "df1 = x1.parse('Sheet1')\n",
    "df1.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Loading the pan genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1458"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myECFile = os.path.join('..','data','Sequencing','EC_Annotation.txt')\n",
    "EC_dict_ori = extractEC(myECFile)\n",
    "len(EC_dict_ori)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extracting all substrates from biolog list with BiGG metabolite identifiers\n",
    "#  for source in df1['BiGG_ID'].astype(str) if source != 'nan']    \n",
    "Substrate_dict = dict()\n",
    "# Substrate_e = dict()\n",
    "# Substrate_c = dict()\n",
    "# Substrate_no = list()\n",
    "for indx, Sub in enumerate(df1['BiGG_ID'].astype(str)):\n",
    "    Sub_e = '{}_e'.format(Sub)\n",
    "    Sub_c = '{}_c'.format(Sub)\n",
    "    Sub_eIdx = np.where([Sub_e==met.id for met in model.metabolites])[0]\n",
    "    Sub_cIdx = np.where([Sub_c==met.id for met in model.metabolites])[0]\n",
    "    CLGrowth = str(round(df1['In_Silico_Growth_Rate'][indx],2))\n",
    "    if Sub_cIdx.size>0 and Sub_eIdx.size>0:\n",
    "        Substrate_dict[Sub] = {'name': model.metabolites[Sub_cIdx[0]].name, 'id_c': model.metabolites[Sub_cIdx[0]].id, 'id_e': model.metabolites[Sub_eIdx[0]].id, 'CL_Growth': CLGrowth}\n",
    "    if Sub_cIdx.size>0 and Sub_eIdx.size<1:\n",
    "        Substrate_dict[Sub] = {'name': model.metabolites[Sub_cIdx[0]].name, 'id_c': model.metabolites[Sub_cIdx[0]].id, 'id_e': None, 'CL_Growth': CLGrowth}\n",
    "    if Sub_cIdx.size<1 and Sub_eIdx.size>0:\n",
    "        Substrate_dict[Sub] = {'name': model.metabolites[Sub_eIdx[0]].name, 'id_c': None, 'id_e': model.metabolites[Sub_eIdx[0]].id, 'CL_Growth': CLGrowth}\n",
    "    if Sub_cIdx.size<1 and Sub_eIdx.size<1:\n",
    "        Substrate_dict[Sub] = {'name': None, 'id_c': None, 'id_e': None}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#  change for loop to new substrate_dict with if loops to check which metabolic compartments exist\n",
    "\n",
    "for Substrate in Substrate_dict.keys():\n",
    "    if Substrate_dict[Substrate]['id_e'] is not None:\n",
    "        EX_Sub = 'EX_{}'.format(Substrate_dict[Substrate]['id_e'])\n",
    "        try: \n",
    "            model.reactions.get_by_id(EX_Sub)\n",
    "        except KeyError:\n",
    "            with model:\n",
    "                model.add_boundary(model.metabolites.get_by_id(Substrate_dict[Substrate]['id_e']), type='exchange')\n",
    "                myGrowth = TestSubstrate(model, EX_Sub)\n",
    "                Substrate_dict[Substrate].update({'growth': str(myGrowth)})\n",
    "#                 print('New exchange reaction for {} with growth: {}'.format(Substrate, myGrowth))        \n",
    "        else:\n",
    "            myGrowth = TestSubstrate(model, EX_Sub)\n",
    "            Substrate_dict[Substrate].update({'growth': str(myGrowth)})\n",
    "#             print('{} with growth: {}'.format(Substrate, myGrowth))\n",
    "    elif Substrate_dict[Substrate]['id_c'] is not None:\n",
    "        myMet = model.metabolites.get_by_id(Substrate_dict[Substrate]['id_c'])\n",
    "        myTransport = CreateTransReact(myMet)\n",
    "        EX_Sub = 'EX_{}_e'.format(Substrate)\n",
    "        with model:\n",
    "            model.add_reactions([myTransport])\n",
    "            model.add_boundary(model.metabolites.get_by_id('{}_e'.format(Substrate)), type='exchange')\n",
    "            myGrowth = TestSubstrate(model, EX_Sub)\n",
    "            Substrate_dict[Substrate].update({'growth': str(myGrowth)})\n",
    "#             print('New transport+exchange reaction for {} with growth: {}'.format(Substrate, myGrowth))        \n",
    "#     else:\n",
    "#         print('{} not in model.'.format(Substrate))\n",
    "        \n",
    "CSVexport(Substrate_dict)\n",
    "print('Biolog-Model growth comparison exported.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "__init__() got an unexpected keyword argument 'nrows'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-182-a0d53c126b14>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Export of gene list with locus tags\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mGenes_File\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'..'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'data'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'Uma_Genes.csv'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mGenes\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mid\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mid\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgenes\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreactions\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnrows\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m400\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'locus-tag'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'name'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0;31m# Genes.to_csv(Genes_File, index=False)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;31m# model.genes.get_by_id('UMAG_00518')\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: __init__() got an unexpected keyword argument 'nrows'"
     ]
    }
   ],
   "source": [
    "# Export of gene list with locus tags\n",
    "Genes_File = os.path.join('..','data','Uma_Genes.csv')\n",
    "Genes = pd.DataFrame([[x.id,i.id] for x in model.genes for i in x.reactions], nrows=400, columns=['locus-tag', 'name'])\n",
    "# Genes.to_csv(Genes_File, index=False)\n",
    "# model.genes.get_by_id('UMAG_00518')\n",
    "# model.reactions.get_by_id('ANTPPT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "iUma in Pan: 513\n",
      "iUma only: 301\n",
      "Pan only: 945\n"
     ]
    }
   ],
   "source": [
    "# EC_dict[model.genes.get_by_id('UMAG_00518').name]\n",
    "EC_dict = EC_dict_ori.copy()\n",
    "Gene_common = [True if ModGene.name in EC_dict.keys() else False for ModGene in model.genes]\n",
    "iUmaANDpan = [myGene.name for myGene in model.genes[Gene_common]]\n",
    "iUmaNOTpan = [myGene.name for myGene in model.genes[list(map(operator.not_, Gene_common))]]\n",
    "[EC_dict.pop(myGene.name, None) for myGene in model.genes]\n",
    "panNOTiUma = list(EC_dict.keys())\n",
    "f1 = os.path.join('..','data','Sequencing', 'Analysis', 'iUmaANDpan.txt')\n",
    "f2 = os.path.join('..','data','Sequencing', 'Analysis','iUmaNOTpan.txt')\n",
    "f3 = os.path.join('..','data','Sequencing', 'Analysis','panNOTiUma.txt')\n",
    "with open(f1, \"w\") as f:\n",
    "    f.write(\"\\n\".join(iUmaANDpan))\n",
    "with open(f2, \"w\") as f:\n",
    "    f.write(\"\\n\".join(iUmaNOTpan))\n",
    "with open(f3, \"w\") as f:\n",
    "    f.write(\"\\n\".join(panNOTiUma))\n",
    "print('iUma in Pan:', sum(Gene_common))\n",
    "print('iUma only:', len(model.genes)-sum(Gene_common))\n",
    "print('Pan only:', len(panNOTiUma))\n",
    "# model.genes.get_by_id('UMAG_00037')\n",
    "# model.reactions.get_by_id('1.6.5.4-RXN')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Unique usti 521 genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 289,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-289-a8968bb9d96d>:41: UserWarning: FixedFormatter should only be used together with FixedLocator\n",
      "  ax2.set_xticklabels(['','512 only','+1','+2','+3','all strains'])\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX0AAAEKCAYAAAD+XoUoAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8/fFQqAAAACXBIWXMAAAsTAAALEwEAmpwYAAAZsUlEQVR4nO3df5gdVZ3n8feHABoiSjCt/EjGME6EDQh5oDfIjDJBMCS4Y0QZhXEhKG5khgzDPPq4UXaTCLobl3WcB6LMRM0GZmIYHUUZE0mCIxNlQNKJHZKISDYGk4CkMQyQxUED3/2jTkPR3Nu3b6dv3abP5/U89+mqU6eqzrmdfLpu3apTigjMzCwPB7W7AWZmVh2HvplZRhz6ZmYZceibmWXEoW9mlhGHvplZRhqGvqSlkvZI2lIqWyhpt6Tu9DqvtOwTkrZJekDSuaXyGalsm6R5Q98VMzNrRI2u05d0JrAPuDkiTkplC4F9EfG/+9SdDKwApgLHAHcAb0qLfwa8A9gFrAcuioifDFlPzMysoYMbVYiIdZImDnB7s4BbIuIZ4OeStlH8AQDYFhHbASTdkuo69M3MKtQw9PsxV9IlQBfw0Yh4HDgWuKdUZ1cqA9jZp/z0WhuVNAeYAzBmzJjTTjjhhANooplZfjZs2PBYRHTUWjbY0L8RuBaI9PNzwIcGua0XiYglwBKAzs7O6OrqGorNmpllQ9JD9ZYNKvQj4tHSxr8EfCfN7gYmlKqOT2X0U25mZhUZ1CWbko4uzZ4P9F7ZcxtwoaRXSDoOmATcS/HF7SRJx0k6FLgw1TUzswo1PNKXtAKYBoyTtAtYAEyTNIXi9M4O4CMAEbFV0tcovqDdD1wREc+m7cwFVgOjgKURsXWoO2NmZv1reMlmO/mcvplZ8yRtiIjOWst8R66ZWUYc+mZmGXHom5llxKFvZpYRh76ZWUYc+mZmGXHom5llxKFvZpYRh76ZWUYc+mZmGXHom5llxKFvZpYRh76ZWUYc+mZmGXHom5llxKFvZpaRhqEvaamkPZK21Fj2UUkhaVyanybpCUnd6TW/VHeGpAckbZM0b2i7YWZmAzGQI/1lwIy+hZImANOBX/RZ9IOImJJe16S6o4AvADOBycBFkiYfSMPNzKx5DUM/ItYBe2ss+jzwcYrn5DYyFdgWEdsj4jfALcCsZhpqZmYHblDn9CXNAnZHxKYai8+QtEnSdyWdmMqOBXaW6uxKZbW2PUdSl6Sunp6ewTTPzMzqaDr0JR0GfBKYX2PxRuANEXEKcAPwrWa3HxFLIqIzIjo7OjqaXd3MzPoxmCP9NwLHAZsk7QDGAxslHRURT0bEPoCIWAUckr7k3Q1MKG1jfCozM7MKHdzsChGxGXhd73wK/s6IeEzSUcCjERGSplL8UfkV8G/AJEnHUYT9hcCfHHjzzcysGQO5ZHMFcDdwvKRdki7rp/oFwBZJm4DrgQujsB+YC6wG7ge+FhFbD7z5ZmbWDEUM5OKb9ujs7Iyurq52N8PM7GVF0oaI6Ky1zHfkmpllxKFvZpYRh76ZWUYc+mZmGXHom5llxKFvZpaRpm/OMjMbribOW9nuJgyZHYve2ZLt+kjfzCwjDn0zs4w49M3MMuLQNzPLiEPfzCwjDn0zs4w49M3MMuLQNzPLiEPfzCwjA3ly1lJJeyRtqbHso5IiPQcXFa6XtE3SfZJOLdWdLenB9Jo9tN0wM7OBGMiR/jJgRt9CSROA6cAvSsUzgUnpNQe4MdU9ElgAnA5MBRZIGnsgDTczs+Y1DP2IWAfsrbHo88DHgfLzFmcBN6fn4t4DHCHpaOBcYG1E7I2Ix4G11PhDYmZmrTWoc/qSZgG7I2JTn0XHAjtL87tSWb3yWtueI6lLUldPT89gmmdmZnU0HfqSDgM+Ccwf+uZARCyJiM6I6Ozo6GjFLszMsjWYI/03AscBmyTtAMYDGyUdBewGJpTqjk9l9crNzKxCTYd+RGyOiNdFxMSImEhxqubUiPglcBtwSbqK5y3AExHxCLAamC5pbPoCd3oqMzOzCg3kks0VwN3A8ZJ2Sbqsn+qrgO3ANuBLwJ8BRMRe4FpgfXpdk8rMzKxCDZ+cFREXNVg+sTQdwBV16i0FljbZPjMzG0K+I9fMLCMOfTOzjDj0zcwy4tA3M8uIQ9/MLCMOfTOzjDj0zcwy4tA3M8uIQ9/MLCMOfTOzjDj0zcwy4tA3M8uIQ9/MLCMOfTOzjDj0zcwyMpCHqCyVtEfSllLZtZLuk9QtaY2kY1L5NElPpPJuSfNL68yQ9ICkbZLmtaY7ZmbWn4Ec6S8DZvQpuy4iTo6IKcB3ePFD0n8QEVPS6xoASaOALwAzgcnARZImH2jjzcysOQ1DPyLWAXv7lD1Zmh0DRIPNTAW2RcT2iPgNcAswq8m2mpnZARr0OX1Jn5G0E/gALz7SP0PSJknflXRiKjsW2FmqsyuV1druHEldkrp6enoG2zwzM6th0KEfEVdHxARgOTA3FW8E3hARpwA3AN8axHaXRERnRHR2dHQMtnlmZlbDUFy9sxx4LxSnfSJiX5peBRwiaRywG5hQWmd8KjMzswoNKvQlTSrNzgJ+msqPkqQ0PTVt/1fAemCSpOMkHQpcCNx2IA03M7PmHdyogqQVwDRgnKRdwALgPEnHA88BDwGXp+oXAH8qaT/wa+DCiAhgv6S5wGpgFLA0IrYOdWfMzKx/DUM/Ii6qUfyVOnUXA4vrLFsFrGqqdWZmNqR8R66ZWUYc+mZmGXHom5llxKFvZpYRh76ZWUYc+mZmGXHom5llxKFvZpYRh76ZWUYc+mZmGXHom5llxKFvZpYRh76ZWUYc+mZmGXHom5llxKFvZpaRhqEvaamkPZK2lMqulXSfpG5JayQdk8ol6XpJ29LyU0vrzJb0YHrNbk13zMysPwM50l8GzOhTdl1EnBwRU4DvAPNT+UxgUnrNAW4EkHQkxWMWTwemAgskjT3QxpuZWXMahn5ErAP29il7sjQ7Bog0PQu4OQr3AEdIOho4F1gbEXsj4nFgLS/9Q2JmZi3W8Bm59Uj6DHAJ8ARwVio+FthZqrYrldUrr7XdORSfEgC2AicNto1VkDQnIpa0ux3tkHPfIe/+D9e+71j0zkr2M1z7X3J7vQWD/iI3Iq6OiAnAcmDuYLdTY7tLIqIzvYZ14CdzGlcZsXLuO+Td/5z7DsO8/xFR90zKUFy9sxx4b5reDUwoLRufyuqVm5lZhQYV+pImlWZnAT9N07cBl6SreN4CPBERjwCrgemSxqYvcKenMjMzq1DDc/qSVgDTgHGSdlFchXOepOOB54CHgMtT9VXAecA24GnggwARsVfStcD6VO+aiHjRl8MvY8P5vF6r5dx3yLv/OfcdXsb9V0Q0rmVmZiOC78g1M8uIQ9/MLCMOfTOzjDj0zcwy4tA3M8uIQ9/MLCMOfTOzjDj0zcwy4tA3M8uIQ9/MLCMOfTOzjLQk9CVNkPR9ST+RtFXSX6TyIyWtTc/JXetHJpqZVaslA66lRyQeHREbJR0ObADeDVwK7I2IRZLmAWMj4r8OeQPMzKymlhzpR8QjEbExTT8F3E/xeMRZwE2p2k0UfwjMzKwiLR9aWdJEYB3Fs25/ERFHpHIBj/fOl+o//4zcyZMnn7Z169aWts/MRo6J81a2uwlD5kCe9ytpdb1HJrb0i1xJrwK+AVwVEU+Wl0Xx1+Ylf3HKz8gdPXp0K5tnZjZSjau3oGWhL+kQisBfHhHfTMWPpvP9vef997Rq/2Zm9lKtunpHwFeA+yPir0qLbgNmp+nZwLdbsX8zM6ut4TNyB+kPgIuBzZK6U9kngUXA1yRdRvFs3fe1aP9mZlZDS0I/In4IqM7is1uxTzMza8x35JqZZcShb2aWEYe+mVlGHPpmZhlp1dU7w8JIuTvvQO7MMzMr85G+mVlGHPpmZhlx6JuZZcShb2aWEYe+mVlGHPpmZhlx6JuZZcShb2aWEYe+mVlGHPpmZhlx6JuZZaRVj0tcKmmPpC2lsoWSdkvqTq/zWrFvMzOrr1VH+suAGTXKPx8RU9JrVYv2bWZmdbQk9CNiHbC3Fds2M7PBq/qc/lxJ96XTP2NrVZA0R1KXpK6enp6Km2dmNrJVGfo3Am8EpgCPAJ+rVSkilkREZ0R0dnR0VNg8M7ORr7LQj4hHI+LZiHgO+BIwtap9m5lZobLQl3R0afZ8YEu9umZm1hoteVyipBXANGCcpF3AAmCapClAADuAj7Ri32ZmVl9LQj8iLqpR/JVW7MvMzAbOd+SamWXEoW9mlhGHvplZRhz6ZmYZceibmWXEoW9mlhGHvplZRhz6ZmYZceibmWXEoW9mlhGHvplZRhz6ZmYZceibmWXEoW9mlhGHvplZRloS+unB53skbSmVHSlpraQH08+aD0Y3M7PWadWR/jJgRp+yecD3ImIS8L00b2ZmFWpJ6EfEOmBvn+JZwE1p+ibg3a3Yt5mZ1VflOf3XR8QjafqXwOtrVZI0R1KXpK6enp7qWmdmloG2fJEbEUHxgPRay5ZERGdEdHZ0dFTcMjOzka3K0H9U0tEA6eeeCvdtZmZUG/q3AbPT9Gzg2xXu28zMaN0lmyuAu4HjJe2SdBmwCHiHpAeBc9K8mZlV6OBWbDQiLqqz6OxW7M/MzAbGd+SamWXEoW9mlhGHvplZRhz6ZmYZceibmWXEoW9mlhGHvplZRhz6ZmYZceibmWWkJXfkmln7TJy3st1NGDI7Fr2z3U0YcXykb2aWEYe+mVlGHPpmZhlx6JuZZcShb2aWEYe+mVlGKr9kU9IO4CngWWB/RHRW3QYzs1y16zr9syLisTbt28wsWz69Y2aWkXaEfgBrJG2QNKfvQklzJHVJ6urp6WlD88zMRq52hP5bI+JUYCZwhaQzywsjYklEdEZEZ0dHRxuaZ2Y2clUe+hGxO/3cA9wKTK26DWZmuao09CWNkXR47zQwHdhSZRvMzHJW9dU7rwduldS7769GxO0Vt8HMLFuVhn5EbAdOqXKflh8PLWxWny/ZNDPLiEPfzCwjDn0zs4w49M3MMuLQNzPLiEPfzCwjDn0zs4w49M3MMuLQNzPLSLseomIt5rtSzawWH+mbmWXEoW9mlhGHvplZRhz6ZmYZceibmWXEoW9mlpHKQ1/SDEkPSNomaV7V+zczy1nVz8gdBXwBmAlMBi6SNLnKNpiZ5azqI/2pwLaI2B4RvwFuAWZV3AYzs2wpIqrbmXQBMCMiPpzmLwZOj4i5pTpzgDlp9njggcoaODjjgMfa3Yg2ybnvkHf/c+47DP/+vyEiOmotGHbDMETEEmBJu9sxUJK6IqKz3e1oh5z7Dnn3P+e+w8u7/1Wf3tkNTCjNj09lZmZWgapDfz0wSdJxkg4FLgRuq7gNZmbZqvT0TkTslzQXWA2MApZGxNYq29ACL5tTUS2Qc98h7/7n3Hd4Gfe/0i9yzcysvXxHrplZRhz6ZmYZySr0Je2QtFlSt6SuUvkfS9oq6TlJnaXyd0jakNbZIOntQ9SOSyUtHopttVu99y4Xkq6T9FNJ90m6VdIR7W5TVSRdm/rdLWmNpGPa3aZGUgaMS9P7mljvKkmHDWJ/10g6p9n1Wimr0E/Oiogpfa6x3QK8B1jXp+5jwB9FxJuB2cDfVdTGYUnSNEnL+hTXe+9GnDr9XwucFBEnAz8DPlF5wypQp+/XRcTJETEF+A4wv/KGVecqoGbop+FlaoqI+RFxR6saNRg5hv5LRMT9EfGSO38j4scR8XCa3QqMlvSKvvUknS3px+kTwdLeOumo4lOSNqZlJ/RZ73BJP5d0SJp/dXn+5aDee5eLiFgTEfvT7D0U955kISKeLM2OAYbNVSGSvpU+nW9Nd/kPdL0xklZK2iRpi6T3S7oSOAb4vqTvp3r7JH1O0ibgDEnzJa1P6yyRpFRvWRqJoG4eSPrD9GmpO+XI4UP+hpTkFvoBrEn/GAb8DyF5L7AxIp4pF0p6JbAMeH/6RHAw8KelKo9FxKnAjcDHXtSYiKeAO4HeJ39fCHwzIn7bZNtsePgQ8N12N6JKkj4jaSfwAYbXkf6HIuI0oBO4UtJrB7jeDODhiDglIk4Cbo+I64GHKc4SnJXqjQF+lOr9EFgcEf8xrTMa+E91tl8rDz4GXJE+Mb0N+HVzXW1ObqH/1vSGzwSukHTmQFaSdCLwWeAjNRYfD/w8In6W5m8Cytv9Zvq5AZhYY/0vAx9M0x8E/s9A2lQlST+S1E3R1neVjkrObXPTKjGQ/ku6GtgPLG9TM1uiUd8j4uqImEDR77n9bKpqV6aj8HsoRgGYNMD1NgPvkPRZSW+LiCfq1HsW+EZp/qz0Xm0G3g6cWGe9WnlwF/BX6RPFEaVPji0x7MbeaaWI2J1+7pF0K8Won/2ei5Y0HrgVuCQi/u8gdtv7yeBZarzfEXGXpImSpgGjImLLIPbRUhFxOhTndYFLI+LSdranao36L+lSiiO7s2OE3fjSxO9+ObAKWFBJw/qR2noOcEZEPC3pTuCVA1k3In4m6VTgPODTkr4XEdfUqPrvEfFs2t8rgS8CnRGxU9LCfvb3kjyIiEWSVqZ93iXp3Ij46UDaOxjZHOmnc3WH904D0ym+hOxvnSOAlcC8iLirTrUHgImSfi/NXwz8S5PNuxn4KsPwKN/6J2kG8HHgXRHxdLvbUyVJ5aPnWUDLgqpJrwEeT4F/AvCWga6YrkB6OiL+HrgOODUtegqod669N+Afk/Qq4IJmGivpjRGxOSI+SzFUzQmN1jkQ2YQ+8Hrgh+kj373Ayoi4HUDS+ZJ2AWcAKyWtTuvMBX4PmF/6WPu68kYj4t8pTst8PX20ew74mybbthwYC6wYZN/app/3LheLKcJgbfr30ezv/uVsUfri8j6Kg6i/aHeDktuBgyXdDyyiOMUzUG8G7k2ntBYAn07lS4Dbe7/ILYuIfwO+RHEQuZoiuJtxVel9/C0t/l7IwzAMA+nb/VkRcXG722JmI1tW5/SHI0k3UHyxfF6722JmI5+P9M3MMpLTOX0zs+w59M3MMuLQNzPLiEN/BJL0bOkS025J8+rUuyRdKrY5jfnxsVr1qqRiYK/fH6p6NdZ7fpRFGzwVI8U2PaqmpMslXdKKNtnA+OqdkenXaRyPuiTNpBg5cHpEPKxikLjh8J9xGrAP+NchqmetcSnFdekP910gaVTv3ap9RURO9zEMTxHh1wh7AfsGUGcd8PYmtzsR+AGwMb1+P5VPoxg47h8p7spczgtXhu0APpXqbwZOSOVHAt8C7qO4eebktP1fAruBborBp/4I+BHwY+AOipvsatXroBgLZX16/UHaz2uBNRSjpH4ZeAgYV6Nvl1EMjXwvxY02i1N5ve0uBJamfm8Hrixt6z+n7XQDf0vxPOhRFAPzbUnvw1/WaMNL+lra103pvX+IYijr/5W2cztwSKp3dlp3c2rbKxr8Djoohoau+97UajfFHaf7KO5G76YYYGwHxfhUGykGDvwv6f3alN6/w0p9+ViavjOtc29679+Wyk8svX/3AZPa/X9qJL3a3gC/WvBLLcb16C693l+jzl7gNU1u9zDglWl6EtCVpqcBT1AMK3wQcDfF4Ha9gfPnafrPgC+n6RuABWn67UB3mn4+FNL8WF74A/Jh4HN16n21tM/fAe5P09cD89P0OylGWu0bbMekdh4JHEIRrosbbHchxaeMVwDjgF+ldf8D8E+8EMRfpPgEdRqwtrTPI2q8v/319Ydp+6cATwMz07JbgXdTDAWwE3hTKr8ZuKrB72Ax8Ik0PaPOe1Oz3RSB3Vkq3wF8vDT/2tL0p0v7f/73lrbR28fzgDtK/zY+kKYPBUa3+//USHr59M7I1PD0ziAdAiyWNIXiD8ubSsvujYhdAOkW9okUQQUvHlnwPWn6rRTDVRMR/yzptZJeXWOf44F/kHQ0RQD8vE7bzgEmp2HMAV6dxkE5s3efEbFS0uM11p0K/EtE7E3t/3qpb/W2C8VQHs8Az0jaQ/Ep5GyKoFyf1hkN7KH4Q/C76Wa8lRSfPprp63cj4rdpqI9RFEf4UBx9T6T2aK9XAH+d5uv9Ds5P783tdd6b7QNod69/KE2fJOnTwBHAqyiGJ6il1qiTdwNXp8EOvxkRD/azT2uSv8jN11aKcGrGXwKPUhxtdlIEU6/ycwb6jija70ijDdxAcdT9ZoqhreuNXngQ8JYonoo2JSKOjYgBPw6vH/1tt1afBdxUqn98RCyMiMcp3rc7gcspTqf01V9fnwGIiOeA30Y6DKYY62kg7+mgfgcDbHev/1eaXgbMTX35FM2NOvlV4F0U48qv0hA9ptQKDv18/U/gOklHAUg6VNKHG6zzGuCRFDwXUxxxDtYPKB680TsU7mNRPImp72iGr6E4dw/FIyt79a23Bvjz3pn0aQSK7y7+JJXNpDiF0td64A8ljZV0MOkTSIPt1vM94ILegfkkHSnpDemKoYMi4hvAf+OF0RvL6vV1IAYz2utdwPtSO6dT473pp939jTpJWvaIiqfAfWCgnUj7/F1gexQPL/k2xfc9NkQc+iPT6D6XbC6C5x/S/C6AiFhFcU73DklbKb6Ae3Wqd7mky2ts94vA7DRS6Qm8+MiuWQuB09LIgot4IeT+CTg/tfttqd7XJW2geGYxdepdCXSqeFD3TyiOSqE4yjwz9fE9wC/6NiSK5yz8D4ovD++iOD/d+/CMetutKSJ+QhGOa1Lf1gJHA8cCd6ZTX39P7Wfp1utrQzG40V4/BUyXtAX4Y4ovx5/qU6deu5cBf5Pe/9E1tv3fKb6Uvovmh1x+H7Al7fMkiu8nbIh47B0zQNKrImJfOtK/FVgaEbe2u12tlC7TfTYi9ks6A7ixRd8F2TDiL3LNCgslnUNx7nkNxeWkI93vAF+TdBDwG4rLLG2E85G+mVlGfE7fzCwjDn0zs4w49M3MMuLQNzPLiEPfzCwj/x8wdMDaD7ID9wAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# loading xls with all data\n",
    "myseqsXls = os.path.join('..','data','Sequencing','Pan_Uma.xlsx')\n",
    "MutualFig = os.path.join('..','data','Sequencing', 'Analysis','Hist_ECGeneOverlap512.svg')\n",
    "\n",
    "\n",
    "myseqs = pd.read_excel(myseqsXls, skiprows=1, index_col=None, nrows=6779, false_values=' - , -') # There are only 6779 positions with U. maydis 521 genes\n",
    "myseqs.replace(' - , -', False, regex=True, inplace=True)\n",
    "\n",
    "mylist = list(myseqs['Usti_521'])\n",
    "\n",
    "# Finding positions in the xls that are also annotated with an ec number\n",
    "IndxKeep = [mylist.index(elem) for elem in mylist for gene in EC_dict_ori if gene in elem]\n",
    "IndxRmve = np.ones(len(mylist), dtype=bool)\n",
    "IndxRmve[IndxKeep] = False\n",
    "myseqs.drop(np.arange(len(mylist))[IndxRmve], inplace=True)\n",
    "\n",
    "# finding the distribution of genes in other strains\n",
    "GeneShare = list()\n",
    "for idx, row in myseqs.iterrows():\n",
    "    GeneShare.append(sum(1 for entry in row if isinstance(entry, str)))\n",
    "    \n",
    "    \n",
    "# histogram with broken y-axis:\n",
    "# https://stackoverflow.com/questions/60062664/y-xis-break-on-matplotlib-histogram\n",
    "\n",
    "bin_edges = np.linspace(.5, 5.5, 6) # make the bins\n",
    "bin_centres = np.linspace(1, 5, 5) # for plotting only\n",
    "my_hist = np.histogram(GeneShare, bins = bin_edges)[0]\n",
    "\n",
    "\n",
    "f, (ax, ax2) = plt.subplots(2,1,sharex = True) # make the axes\n",
    "ax.bar(bin_centres, my_hist) # plot on top axes\n",
    "ax2.bar(bin_centres, my_hist) # plot on bottom axes\n",
    "ax.set_ylim([1300,1500]) # numbers here are specific to this example\n",
    "ax2.set_ylim([0, 20]) # numbers here are specific to this example\n",
    "ax.spines['bottom'].set_visible(False)\n",
    "ax2.spines['top'].set_visible(False)\n",
    "ax.xaxis.tick_bottom() \n",
    "ax.tick_params(labeltop=False)\n",
    "ax2.xaxis.tick_bottom()\n",
    "ax2.set_xticklabels(['','512 only','+1','+2','+3','all strains'])\n",
    "plt.xlabel('E.C. annotated genes among strains')\n",
    "plt.savefig(MutualFig)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extracting the 512 unique E.C. genes and export to fasta file for KAAS analysis\n",
    "ECUniqFile = os.path.join('..','data','Sequencing', 'Analysis','ECUnique512.fasta')\n",
    "\n",
    "ECUniq = [elem.split(',')[0] for elem in myseqs.iloc[np.arange(len(GeneShare))[np.array(GeneShare)==1]]['Usti_521'].values]\n",
    "\n",
    "myFasta = 'pangenome_AA_Uma.faa'\n",
    "myDir = os.path.join('..','data','Sequencing')\n",
    "myFile = os.path.join(myDir,myFasta)\n",
    "\n",
    "records = list()\n",
    "with open(myFile) as infile:\n",
    "    for record in SeqIO.parse(infile, 'fasta'):\n",
    "        if record.id in ECUniq:\n",
    "            records.append(record)\n",
    "with open(os.path.join(ECUniqFile), 'w') as outfile:\n",
    "    SeqIO.write(records, outfile, \"fasta\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### export of panNOTiUma list as fasta for KAAS analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "myFasta = 'pangenome_AA_Uma.faa'\n",
    "myDir = os.path.join('..','data','Sequencing')\n",
    "myFile = os.path.join(myDir,myFasta)\n",
    "\n",
    "records = list()\n",
    "with open(myFile) as infile:\n",
    "    for record in SeqIO.parse(infile, 'fasta'):\n",
    "        if record.id in panNOTiUsti:\n",
    "            records.append(record)\n",
    "with open(os.path.join(myDir,'panNOTiUsti.fasta'), 'w') as outfile:\n",
    "    SeqIO.write(records, outfile, \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Testing KEGG Biopython functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.KEGG import REST\n",
    "result = REST.kegg_get('ko:k03800').read()\n",
    "BRidx = result.find('BR:ko')\n",
    "GOidx = result.find('map')\n",
    "BRid = result[BRidx+5:BRidx+10]\n",
    "GOid = result[GOidx+3:GOidx+8]\n",
    "print(BRid, GOid)\n",
    "print(result)\n",
    "\n",
    "# finding multiple 'map' id in string of result\n",
    "# https://stackoverflow.com/questions/47486563/how-can-i-get-index-of-two-of-more-duplicate-characters-in-a-string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "KAAStxt = 'KaasMap_panNOTiUsti.txt'\n",
    "KAASpath = os.path.join('..','data','Sequencing','Analysis', KAAStxt)\n",
    "myKAAS = pd.read_csv(KAASpath, delimiter='\\t', names=['UMAG-ID','KO-ID'])\n",
    "myKAAS\n",
    "MAP_dict = dict()\n",
    "for ko in myKAAS['KO-ID']:\n",
    "    if str(ko) != 'nan':\n",
    "        result = REST.kegg_get('ko:{}'.format(ko)).read()\n",
    "        MAPidx = result.find('map')\n",
    "        MAPid = result[MAPidx+3:MAPidx+8]\n",
    "        if MAPid in MAP_dict:\n",
    "            MAP_dict[MAPid]+=1\n",
    "        else:\n",
    "            MAP_dict[MAPid] = 1\n",
    "MAP_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function definitions\n",
    "\n",
    "We are testing three different growth conditions with substrates methanol, glycerol, and glucose. For each of the substrates, the corresponding exchange rate has to be activated, the approriate biomass composition formula has to be selected and exchange reactions for non-used substrates need to be closed."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simulation\n",
    "\n",
    "Here, we start the simulation loop. For all substrates we use the substrate uptake rate and use the functions defined previously to update the model. The final growth rate is stored, and is then used in a plot versus the experimentally observed growth."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.reactions.query('ATP')\n",
    "model.reactions.ATPS\n",
    "# model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# checking TCA consistency with objective to ATP hydrolysis/maintenance\n",
    "with model as ATPM:\n",
    "    ATPM.objective = 'ATPS'\n",
    "    solution = ATPM.optimize()\n",
    "    print(solution.fluxes['BIOMASS_REACTION'])\n",
    "    print(model.summary())\n",
    "    \n",
    "# visualization    \n",
    "builder = Builder()\n",
    "# load U. maydis ESCHER map with Pentose phosphate pathway, Glycolysis\n",
    "\n",
    "Escher_Central = os.path.join('Maps','umay_tcav7.json')\n",
    "Escher_Glycine = os.path.join('Maps','umay_glycinev2.json')\n",
    "builder = Builder(\n",
    "    map_json=Escher_Central,\n",
    "    model = model, # 'iCL1079.json',\n",
    ")\n",
    "# Run FBA with the model and add the flux data to the map\n",
    "solution = builder.model.optimize()\n",
    "builder.reaction_data = solution.fluxes\n",
    "\n",
    "builder.save_html('example_map.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparison of BIOLOG and model\n",
    "\n",
    "### Loading substrate list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.metabolites.get_by_id('gdpfuc_c')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.metabolites.query('Glucose')\n",
    "myMets = pd.DataFrame([[x.name,x.id] for x in model.metabolites])\n",
    "#     print('%9s : %s' % (x.id, x.formula))\n",
    "# myMets\n",
    "check = 'fuc'\n",
    "matching = [idx for idx,s in enumerate(myMets[1]) if check in s]\n",
    "myMets.loc[matching]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model_loop = model.copy()\n",
    "growth_simulated = [];\n",
    "for index, row in df1.iterrows():\n",
    "  print(index)\n",
    "  if row['Substrate'] == 'Methanol':\n",
    "    model_loop = AdaptMethanol(model_loop, row['uptake rate (mmol/gCDW/h)'])\n",
    "  elif row['Substrate'] == 'Glucose':\n",
    "    model_loop = AdaptGlucose(model_loop, row['uptake rate (mmol/gCDW/h)'])\n",
    "  elif row['Substrate'] == 'Glycerol':\n",
    "    model_loop = AdaptGlycerol(model_loop, row['uptake rate (mmol/gCDW/h)'])\n",
    "  else:\n",
    "    print('substrate not considered')      \n",
    "  model_loop.optimize()\n",
    "  growth_simulated.append(model_loop.objective.value)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rc('axes', labelsize=12) \n",
    "plt.rc('xtick', labelsize=10) \n",
    "plt.rc('ytick', labelsize=10) \n",
    "plt.rc('legend', fontsize=12) \n",
    "# plt.figure(figsize=(4,3))\n",
    "plt.scatter(df1['growth rate (/h)'][1:8], growth_simulated[1:8], s=100, c='k', marker='o', );\n",
    "plt.scatter(df1['growth rate (/h)'][8], growth_simulated[8], s=100, c='k', marker='s');\n",
    "plt.scatter(df1['growth rate (/h)'][0], growth_simulated[0], s=100, c='k', marker='x');\n",
    "plt.xlabel('Experiment (/h)');\n",
    "plt.ylabel('Simulation (/h)');\n",
    "myline = np.linspace(0,np.max(growth_simulated),10);\n",
    "plt.plot(myline,myline,'k--');\n",
    "# plt.title('Growth rate comparison');\n",
    "# correlation coefficient between experimental and simulated data\n",
    "R2 = np.corrcoef(np.array([growth_simulated[0:9],df1['growth rate (/h)'][0:9]]))\n",
    "R2Legend = r'Optimum, $R^2 = %.4f$'%(R2[0,1]) \n",
    "plt.legend([R2Legend,'Methanol (van Dijken)','Glycerol (deKoning, Moon)','Glucose (Lehnen)'], loc=2);\n",
    "plt.style.use('seaborn-paper')\n",
    "# plt.savefig('iOpol_growth_exp-vs-sim.png', bbox_inches='tight')\n",
    "# plt.savefig('iOpol_growth_exp-vs-sim.pdf', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Memote quality test\n",
    "\n",
    "Memote performs quality assessments of genome scale metabolic models. The process can take a while, around 5 min."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mem_return, mem_result = test_model(model, results=True)\n",
    "with open('iUmay_MemoteReport.html', 'w') as f:\n",
    "    f.write(snapshot_report(mem_result))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## correcting reactions and metabolites for balance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# A-3-OXO-ACID_c is unbalanced\n",
    "DelRct = ['3-OXOACID-COA-TRANSFERASE-RXN']\n",
    "model.remove_reactions(DelRct, remove_orphans=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Objective\n",
      "=========\n",
      "1.0 BIOMASS_REACTION = 0.4939843816958042\n",
      "\n",
      "Uptake\n",
      "------\n",
      "Metabolite    Reaction    Flux  C-Number C-Flux\n",
      "     co2_e    EX_co2_e   2.959         1  4.70%\n",
      "  glc__D_e EX_glc__D_e      10         6 95.30%\n",
      "     nh4_e    EX_nh4_e   2.717         0  0.00%\n",
      "      o2_e     EX_o2_e   3.174         0  0.00%\n",
      "      pi_e     EX_pi_e  0.4887         0  0.00%\n",
      "     so4_e    EX_so4_e 0.04051         0  0.00%\n",
      "\n",
      "Secretion\n",
      "---------\n",
      "Metabolite    Reaction    Flux  C-Number C-Flux\n",
      "     fum_e    EX_fum_e -0.2366         4  2.31%\n",
      "     h2o_e    EX_h2o_e  -13.58         0  0.00%\n",
      "  mal__L_e EX_mal__L_e  -10.02         4 97.69%\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[0.0]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.optimize()\n",
    "print(model.summary())\n",
    "[reaction.flux for reaction in model.reactions.query('TRNA')]\n",
    "# print(model.metabolites.get_by_id('fum[c]').summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "any(['3-OXO-ACID' in mylst.id for mylst in lowMet])\n",
    "# lowMet[0]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dependencies and versions\n",
    "The code below was used to identify the versions of dependent packages."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext watermark\n",
    "%watermark -v -m -p ipywidgets,matplotlib,numpy,pandas,cobra,memote,watermark"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## model changes start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# maltose is used as substrate according to biolog\n",
    "# adding maltase reactions\n",
    "model.reactions.get_by_id('RXN-14350').id = 'Maltase'\n",
    "model.reactions.get_by_id('Maltase').build_reaction_from_string('h2o_c + malt_c --> 2 glc__D_c')\n",
    "# maltose is preferentially metabolized by an unusual polymerization and depolymerization stept via Amalt1\n",
    "model.reactions.get_by_id('AMALT1').bounds = 0,0\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "write_sbml_model(model, 'iUma22_BiologCorr_Maltase.xml')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## model changes stop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Maltase: 10.0 (h2o_c + malt_c --> 2.0 glc__D_c)\n",
      "Trans_malt_c: -10.0 (malt_c <=> malt_e)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[None, None]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TrackMet = 'malt_c' # malt_c\n",
    "medium = model.medium\n",
    "medium['EX_glc__D_e'] = 0 \n",
    "medium['EX_malt_e'] = 10\n",
    "model.medium = medium\n",
    "solution = model.optimize()\n",
    "[print('{}: {} ({})'.format(rct.id,solution[rct.id], rct.reaction)) for rct in model.metabolites.get_by_id(TrackMet).reactions if abs(solution[rct.id])>0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.metabolites.get_by_id('Lfmkynr_c').charge = 2."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "iambjudas",
   "language": "python",
   "name": "iambjudas"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
