{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genome scale model reconstruction of the smut fungus *Ustilago maydis*: Reproduction of simulations with CobraPy Notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction\n",
    "\n",
    "This notebook is an extension to the article by Liebal et al., (2022) and reproduces the comparison of experimental growth rates with predictions by the genome scale metabolic model (GSMM) of *Ustilago maydis*. Finaly, the Memote quality check is performed. \n",
    "\n",
    "## Requirements\n",
    "Make sure that the required files are present in the notebook directory. \n",
    "\n",
    "Files:\n",
    " * ../model/iCL1079.xml: SBML file, Genome scale metabolic model of *U. maydis*\n",
    " \n",
    "Dependencies:\n",
    " * CPython 3.7.6\n",
    " * IPython 7.12.0\n",
    " * ipywidgets 7.5.1\n",
    " * matplotlib 3.1.3\n",
    " * numpy 1.18.1\n",
    " * pandas 1.0.1\n",
    " * cobra 0.17.1\n",
    " * memote 0.10.2\n",
    "\n",
    "## Set-up compute environment \n",
    "\n",
    "In the following cell the python environment is set up. If CobraPy and Memote are not already downloaded, use the following command to download the packages them from the cell:\n",
    "!{sys.executable} -m pip install cobra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys # loading commands to control/navigate within the system architecture\n",
    "# Loading pandas, a library for data manipulation\n",
    "import os\n",
    "# import xlrd\n",
    "import pandas as pd\n",
    "# import lxml\n",
    "# for export of biolog substrate test dictionary\n",
    "import csv\n",
    "import itertools\n",
    "import operator\n",
    "\n",
    "\n",
    "# Loading numpy, a library fo manipulation of numbers\n",
    "import numpy as np\n",
    "\n",
    "# loading matplotlib, a library for visualization\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "# loading cobrapy, a library dedicated to the analysis of genome scale metabolic models\n",
    "from cobra.io import read_sbml_model, write_sbml_model\n",
    "from cobra import Reaction, Metabolite\n",
    "\n",
    "# from escher import Builder\n",
    "\n",
    "# loading Memote, quality assessment of GSMM\n",
    "# from memote import test_model, snapshot_report\n",
    "# from FastaTools import *\n",
    "from iambcodes.cobra import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Support functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def TestSubstrate(model, EX_Sub, Rate=10):\n",
    "    '''\n",
    "    Function for testing substrate consumption with exchange reaction.\n",
    "    '''\n",
    "    with model:\n",
    "        medium = model.medium\n",
    "        medium[EX_Sub] = Rate\n",
    "        medium['EX_co2_e'] = 0\n",
    "        medium['EX_glc__D_e'] = 0\n",
    "        model.medium = medium\n",
    "        return round(model.slim_optimize(),2)\n",
    "\n",
    "def MetNewFromExist(MetIn):\n",
    "    Metout = Metabolite('{}_e'.format(MetIn.id[:-2]),\n",
    "                       formula = MetIn.formula,\n",
    "                       name = MetIn.name,\n",
    "                       compartment = 'e')\n",
    "    return Metout\n",
    "\n",
    "def CreateTransReact(MetIn):\n",
    "    reaction = Reaction('Trans_{}'.format(MetIn.id),\n",
    "                       name = 'Transport c<->e {}'.format(MetIn.name),\n",
    "                       lower_bound = -1000,\n",
    "                       upper_bound = 1000)\n",
    "    MetOut = MetNewFromExist(MetIn)\n",
    "    reaction.add_metabolites({MetIn:-1.0, MetOut:1.0})\n",
    "    \n",
    "    return reaction\n",
    "\n",
    "def CSVexport(myDict, FName='BiologGrowthTest'):\n",
    "    fields = ['id','name','id_e','id_c','growth','CL_Growth']\n",
    "    with open(FName, 'w') as f:\n",
    "        w = csv.DictWriter(f, fields)\n",
    "        w.writeheader()\n",
    "        for k in myDict:\n",
    "            w.writerow({field: myDict[k].get(field) or k for field in fields})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data loading\n",
    "\n",
    "Now, the external data is integrated. The excel-sheet is parsed and the GSMM file is converted to a model variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "myxls = os.path.join('..','data','MSc-Lieven_Christian','sorted_growth_study.xlsx')\n",
    "x1 = pd.ExcelFile(myxls)\n",
    "df1 = x1.parse('Sheet1')\n",
    "df1.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Loading the pan genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "myECFile = os.path.join('..','data','Sequencing','EC_Annotation.txt')\n",
    "EC_dict = extractEC(myECFile)\n",
    "len(EC_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Model loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "ModelFile = os.path.join('..','model','iCL1079.xml') # iCL1079.xml\n",
    "model=read_sbml_model(ModelFile)\n",
    "model.metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.reactions.get_by_id('RXN-12710')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extracting all substrates from biolog list with BiGG metabolite identifiers\n",
    "#  for source in df1['BiGG_ID'].astype(str) if source != 'nan']    \n",
    "Substrate_dict = dict()\n",
    "# Substrate_e = dict()\n",
    "# Substrate_c = dict()\n",
    "# Substrate_no = list()\n",
    "for indx, Sub in enumerate(df1['BiGG_ID'].astype(str)):\n",
    "    Sub_e = '{}_e'.format(Sub)\n",
    "    Sub_c = '{}_c'.format(Sub)\n",
    "    Sub_eIdx = np.where([Sub_e==met.id for met in model.metabolites])[0]\n",
    "    Sub_cIdx = np.where([Sub_c==met.id for met in model.metabolites])[0]\n",
    "    CLGrowth = str(round(df1['In_Silico_Growth_Rate'][indx],2))\n",
    "    if Sub_cIdx.size>0 and Sub_eIdx.size>0:\n",
    "        Substrate_dict[Sub] = {'name': model.metabolites[Sub_cIdx[0]].name, 'id_c': model.metabolites[Sub_cIdx[0]].id, 'id_e': model.metabolites[Sub_eIdx[0]].id, 'CL_Growth': CLGrowth}\n",
    "    if Sub_cIdx.size>0 and Sub_eIdx.size<1:\n",
    "        Substrate_dict[Sub] = {'name': model.metabolites[Sub_cIdx[0]].name, 'id_c': model.metabolites[Sub_cIdx[0]].id, 'id_e': None, 'CL_Growth': CLGrowth}\n",
    "    if Sub_cIdx.size<1 and Sub_eIdx.size>0:\n",
    "        Substrate_dict[Sub] = {'name': model.metabolites[Sub_eIdx[0]].name, 'id_c': None, 'id_e': model.metabolites[Sub_eIdx[0]].id, 'CL_Growth': CLGrowth}\n",
    "    if Sub_cIdx.size<1 and Sub_eIdx.size<1:\n",
    "        Substrate_dict[Sub] = {'name': None, 'id_c': None, 'id_e': None}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(Substrate_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#  change for loop to new substrate_dict with if loops to check which metabolic compartments exist\n",
    "\n",
    "for Substrate in Substrate_dict.keys():\n",
    "    if Substrate_dict[Substrate]['id_e'] is not None:\n",
    "        EX_Sub = 'EX_{}'.format(Substrate_dict[Substrate]['id_e'])\n",
    "        try: \n",
    "            model.reactions.get_by_id(EX_Sub)\n",
    "        except KeyError:\n",
    "            with model:\n",
    "                model.add_boundary(model.metabolites.get_by_id(Substrate_dict[Substrate]['id_e']), type='exchange')\n",
    "                myGrowth = TestSubstrate(model, EX_Sub)\n",
    "                Substrate_dict[Substrate].update({'growth': str(myGrowth)})\n",
    "#                 print('New exchange reaction for {} with growth: {}'.format(Substrate, myGrowth))        \n",
    "        else:\n",
    "            myGrowth = TestSubstrate(model, EX_Sub)\n",
    "            Substrate_dict[Substrate].update({'growth': str(myGrowth)})\n",
    "#             print('{} with growth: {}'.format(Substrate, myGrowth))\n",
    "    elif Substrate_dict[Substrate]['id_c'] is not None:\n",
    "        myMet = model.metabolites.get_by_id(Substrate_dict[Substrate]['id_c'])\n",
    "        myTransport = CreateTransReact(myMet)\n",
    "        EX_Sub = 'EX_{}_e'.format(Substrate)\n",
    "        with model:\n",
    "            model.add_reactions([myTransport])\n",
    "            model.add_boundary(model.metabolites.get_by_id('{}_e'.format(Substrate)), type='exchange')\n",
    "            myGrowth = TestSubstrate(model, EX_Sub)\n",
    "            Substrate_dict[Substrate].update({'growth': str(myGrowth)})\n",
    "#             print('New transport+exchange reaction for {} with growth: {}'.format(Substrate, myGrowth))        \n",
    "#     else:\n",
    "#         print('{} not in model.'.format(Substrate))\n",
    "        \n",
    "CSVexport(Substrate_dict)\n",
    "print('Biolog-Model growth comparison exported.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export of gene list with locus tags\n",
    "Genes_File = os.path.join('..','data','Uma_Genes.csv')\n",
    "Genes = pd.DataFrame([[x.id,i.id] for x in model.genes for i in x.reactions], columns=['locus-tag', 'name'])\n",
    "# Genes.to_csv(Genes_File, index=False)\n",
    "# model.genes.get_by_id('UMAG_00518')\n",
    "# model.reactions.get_by_id('ANTPPT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# EC_dict[model.genes.get_by_id('UMAG_00518').name]\n",
    "Gene_common = [True if ModGene.name in EC_dict.keys() else False for ModGene in model.genes]\n",
    "iUstiANDpan = [myGene.name for myGene in model.genes[Gene_common]]\n",
    "iUstiNOTpan = [myGene.name for myGene in model.genes[list(map(operator.not_, Gene_common))]]\n",
    "[EC_dict.pop(myGene.name, None) for myGene in model.genes]\n",
    "panNOTiUsti = list(EC_dict.keys())\n",
    "f1 = os.path.join('..','data','Sequencing','iUmaANDpan.txt')\n",
    "f2 = os.path.join('..','data','Sequencing','iUmaNOTpan.txt')\n",
    "f3 = os.path.join('..','data','Sequencing','panNOTiUma.txt')\n",
    "with open(f1, \"w\") as f:\n",
    "    f.write(\"\\n\".join(iUmaANDpan))\n",
    "with open(f2, \"w\") as f:\n",
    "    f.write(\"\\n\".join(iUmaNOTpan))\n",
    "with open(f3, \"w\") as f:\n",
    "    f.write(\"\\n\".join(panNOTiUma))\n",
    "print('iUma in Pan:', sum(Gene_common))\n",
    "print('iUma only:', len(model.genes)-sum(Gene_common))\n",
    "print('Pan only:', len(panNOTiUma))\n",
    "# model.genes.get_by_id('UMAG_00037')\n",
    "# model.reactions.get_by_id('1.6.5.4-RXN')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### export of panNOTiUma list as fasta for KAAS analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "myFasta = 'pangenome_AA_Uma.faa'\n",
    "myDir = os.path.join('..','data','Sequencing')\n",
    "myFile = os.path.join(myDir,myFasta)\n",
    "\n",
    "records = list()\n",
    "with open(myFile) as infile:\n",
    "    for record in SeqIO.parse(infile, 'fasta'):\n",
    "        if record.id in panNOTiUsti:\n",
    "            records.append(record)\n",
    "with open(os.path.join(myDir,'panNOTiUsti.fasta'), 'w') as outfile:\n",
    "    SeqIO.write(records, outfile, \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Testing KEGG Biopython functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.KEGG import REST\n",
    "result = REST.kegg_get('ko:k03800').read()\n",
    "BRidx = result.find('BR:ko')\n",
    "GOidx = result.find('map')\n",
    "BRid = result[BRidx+5:BRidx+10]\n",
    "GOid = result[GOidx+3:GOidx+8]\n",
    "print(BRid, GOid)\n",
    "print(result)\n",
    "\n",
    "# finding multiple 'map' id in string of result\n",
    "# https://stackoverflow.com/questions/47486563/how-can-i-get-index-of-two-of-more-duplicate-characters-in-a-string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "KAAStxt = 'KaasMap_panNOTiUsti.txt'\n",
    "KAASpath = os.path.join('..','data','Sequencing','Analysis', KAAStxt)\n",
    "myKAAS = pd.read_csv(KAASpath, delimiter='\\t', names=['UMAG-ID','KO-ID'])\n",
    "myKAAS\n",
    "MAP_dict = dict()\n",
    "for ko in myKAAS['KO-ID']:\n",
    "    if str(ko) != 'nan':\n",
    "        result = REST.kegg_get('ko:{}'.format(ko)).read()\n",
    "        MAPidx = result.find('map')\n",
    "        MAPid = result[MAPidx+3:MAPidx+8]\n",
    "        if MAPid in MAP_dict:\n",
    "            MAP_dict[MAPid]+=1\n",
    "        else:\n",
    "            MAP_dict[MAPid] = 1\n",
    "MAP_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function definitions\n",
    "\n",
    "We are testing three different growth conditions with substrates methanol, glycerol, and glucose. For each of the substrates, the corresponding exchange rate has to be activated, the approriate biomass composition formula has to be selected and exchange reactions for non-used substrates need to be closed."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simulation\n",
    "\n",
    "Here, we start the simulation loop. For all substrates we use the substrate uptake rate and use the functions defined previously to update the model. The final growth rate is stored, and is then used in a plot versus the experimentally observed growth."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.reactions.query('ATP')\n",
    "model.reactions.ATPS\n",
    "# model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# checking TCA consistency with objective to ATP hydrolysis/maintenance\n",
    "with model as ATPM:\n",
    "    ATPM.objective = 'ATPS'\n",
    "    solution = ATPM.optimize()\n",
    "    print(solution.fluxes['BIOMASS_REACTION'])\n",
    "    print(model.summary())\n",
    "    \n",
    "# visualization    \n",
    "builder = Builder()\n",
    "# load U. maydis ESCHER map with Pentose phosphate pathway, Glycolysis\n",
    "\n",
    "Escher_Central = os.path.join('Maps','umay_tcav7.json')\n",
    "Escher_Glycine = os.path.join('Maps','umay_glycinev2.json')\n",
    "builder = Builder(\n",
    "    map_json=Escher_Central,\n",
    "    model = model, # 'iCL1079.json',\n",
    ")\n",
    "# Run FBA with the model and add the flux data to the map\n",
    "solution = builder.model.optimize()\n",
    "builder.reaction_data = solution.fluxes\n",
    "\n",
    "builder.save_html('example_map.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparison of BIOLOG and model\n",
    "\n",
    "### Loading substrate list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.metabolites.get_by_id('gdpfuc_c')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.metabolites.query('Glucose')\n",
    "myMets = pd.DataFrame([[x.name,x.id] for x in model.metabolites])\n",
    "#     print('%9s : %s' % (x.id, x.formula))\n",
    "# myMets\n",
    "check = 'fuc'\n",
    "matching = [idx for idx,s in enumerate(myMets[1]) if check in s]\n",
    "myMets.loc[matching]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model_loop = model.copy()\n",
    "growth_simulated = [];\n",
    "for index, row in df1.iterrows():\n",
    "  print(index)\n",
    "  if row['Substrate'] == 'Methanol':\n",
    "    model_loop = AdaptMethanol(model_loop, row['uptake rate (mmol/gCDW/h)'])\n",
    "  elif row['Substrate'] == 'Glucose':\n",
    "    model_loop = AdaptGlucose(model_loop, row['uptake rate (mmol/gCDW/h)'])\n",
    "  elif row['Substrate'] == 'Glycerol':\n",
    "    model_loop = AdaptGlycerol(model_loop, row['uptake rate (mmol/gCDW/h)'])\n",
    "  else:\n",
    "    print('substrate not considered')      \n",
    "  model_loop.optimize()\n",
    "  growth_simulated.append(model_loop.objective.value)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rc('axes', labelsize=12) \n",
    "plt.rc('xtick', labelsize=10) \n",
    "plt.rc('ytick', labelsize=10) \n",
    "plt.rc('legend', fontsize=12) \n",
    "# plt.figure(figsize=(4,3))\n",
    "plt.scatter(df1['growth rate (/h)'][1:8], growth_simulated[1:8], s=100, c='k', marker='o', );\n",
    "plt.scatter(df1['growth rate (/h)'][8], growth_simulated[8], s=100, c='k', marker='s');\n",
    "plt.scatter(df1['growth rate (/h)'][0], growth_simulated[0], s=100, c='k', marker='x');\n",
    "plt.xlabel('Experiment (/h)');\n",
    "plt.ylabel('Simulation (/h)');\n",
    "myline = np.linspace(0,np.max(growth_simulated),10);\n",
    "plt.plot(myline,myline,'k--');\n",
    "# plt.title('Growth rate comparison');\n",
    "# correlation coefficient between experimental and simulated data\n",
    "R2 = np.corrcoef(np.array([growth_simulated[0:9],df1['growth rate (/h)'][0:9]]))\n",
    "R2Legend = r'Optimum, $R^2 = %.4f$'%(R2[0,1]) \n",
    "plt.legend([R2Legend,'Methanol (van Dijken)','Glycerol (deKoning, Moon)','Glucose (Lehnen)'], loc=2);\n",
    "plt.style.use('seaborn-paper')\n",
    "# plt.savefig('iOpol_growth_exp-vs-sim.png', bbox_inches='tight')\n",
    "# plt.savefig('iOpol_growth_exp-vs-sim.pdf', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Memote quality test\n",
    "\n",
    "Memote performs quality assessments of genome scale metabolic models. The process can take a while, around 5 min."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mem_return, mem_result = test_model(model, results=True)\n",
    "with open('iUmay_MemoteReport.html', 'w') as f:\n",
    "    f.write(snapshot_report(mem_result))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mass balance test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "# myFile = os.path.join('..','iUL_versions','iUL959_200810.xml') # os.path.join('..','iUL_versions','frank','out_metadata.xml') # os.path.join('..','iUL_versions','iUL959_112718.xml')\n",
    "# model = read_sbml_model(myFile)\n",
    "mysol = [[reaction.id, reaction.reaction, reaction.check_mass_balance()] for reaction in model.reactions if reaction.check_mass_balance()]\n",
    "print('unbalanced reactions: {}'.format(len(mysol)))\n",
    "df_mysol = pd.DataFrame(mysol, columns=['Abbreviation','Formula','Misbalance'])\n",
    "\n",
    "# deleting reactions with following strings\n",
    "DelStr_all = ['Ex_', 'avg', 'BIOM', 'PROT', 'LIP', 'STER', 'DNA', 'RNA', 'CARB', 'GLYCE']\n",
    "for DelStr in DelStr_all:\n",
    "    indices = [i for i, s in enumerate(list(df_mysol['Abbreviation'])) if DelStr in s]\n",
    "    df_mysol.drop(df_mysol.index[[indices]], inplace=True)\n",
    "    \n",
    "df_mysol.reset_index(inplace=True)\n",
    "df_mysol.to_csv('iOpol_unbalanced.txt', index=False)\n",
    "\n",
    "df_mysol\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.optimize()\n",
    "print(model.summary())\n",
    "[reaction.flux for reaction in model.reactions.query('TRNA')]\n",
    "# print(model.metabolites.get_by_id('fum[c]').summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.metabolites.query('tre_c')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dependencies and versions\n",
    "The code below was used to identify the versions of dependent packages."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext watermark\n",
    "%watermark -v -m -p ipywidgets,matplotlib,numpy,pandas,cobra,memote,watermark"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "iambjudas",
   "language": "python",
   "name": "iambjudas"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
